src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs

```diff
@@ -2213,18 +2213,613 @@ class C
+                //     public partial RS this[scoped ref int i, int _] { get => default; set { } }
+                Diagnostic(ErrorCode.ERR_ScopedMismatchInParameterOfPartial, "this").WithArguments("i").WithLocation(11, 23),
+                // (11,35): error CS0631: ref and out are not valid in this context
+                //     public partial RS this[scoped ref int i, int _] { get => default; set { } }
+                Diagnostic(ErrorCode.ERR_IllegalRefParam, "ref").WithLocation(11, 35));
+        }
+
+        [Fact]
+        public void Semantics_OptionalParameters()
+        {
+            var source = """
+                using System;
+
+                var c = new C();
+                Console.Write(c[1]);
+                Console.Write(c[1, 2]);
+
+                partial class C
+                {
+                    public partial int this[int x, int y = 1] { get; }
+                    public partial int this[int x, int y] { get => y; }
+                }
+                """;
+
+            var verifier = CompileAndVerify(source, expectedOutput: "12");
+            verifier.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void OptionalParameters_OnImplementationPart()
+        {
+            // A warning is reported for optional parameters on implementation part, even if it matches the definition part.
+            var source = """
+                partial class C
+                {
+                    public partial int this[int x, int y] { get; set; }
+                    public partial int this[int x, int y = 1] { get => y; set { } }
+
+                    public partial int this[int x, int y, int z = 1] { get; set; }
+                    public partial int this[int x, int y, int z = 1] { get => z; set { } }
+                }
+                """;
+
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (4,40): warning CS1066: The default value specified for parameter 'y' will have no effect because it applies to a member that is used in contexts that do not allow optional arguments
+                //     public partial int this[int x, int y = 1] { get => y; set { } }
+                Diagnostic(ErrorCode.WRN_DefaultValueForUnconsumedLocation, "y").WithArguments("y").WithLocation(4, 40),
+                // (7,47): warning CS1066: The default value specified for parameter 'z' will have no effect because it applies to a member that is used in contexts that do not allow optional arguments
+                //     public partial int this[int x, int y, int z = 1] { get => z; set { } }
+                Diagnostic(ErrorCode.WRN_DefaultValueForUnconsumedLocation, "z").WithArguments("z").WithLocation(7, 47));
+        }
+
+        [Fact]
+        public void OptionalParameters_AllParametersAreOptional()
+        {
+            // An indexer access needs at least one argument in order to be valid
```

Comment:
I think we're missing an error here. There are warnings reported on the implementation, and an error on the call-site, but there should also be an error on the definition since that indexer won't be callable.

Reply:
I don't think an error is expected here. A warning is reported when an indexer has a single parameter which is also optional, but no warning is reported when the indexer has multiple parameters, even if they are all optional.

The indexer can still be used when at least one explicit argument is provided, and using labeled arguments, it's possible to exercise either of the default values, e.g. with `this[x: 0]` or `this[y: 0]` in this test.

In this case, warnings are only being reported because default values are present on an implementation part. There is no issue in the definition part.